tutorials/basics/pytorch_basics/main.cpp

```diff
@@ -1,91 +1,244 @@
 
-  std::cout << std::endl << "SAVE AND LOAD THE MODEL: " << std::endl;
+    // Compute loss
+    auto loss = criterion(pred, y);
+    std::cout << "Loss: " << loss.item<double>() << '\n';
+
+    // Backward pass
+    loss.backward();
+
+    // Print out the gradients
+    std::cout << "dL/dw:\n" << linear->weight.grad() << '\n';
+    std::cout << "dL/db:\n" << linear->bias.grad() << '\n';
+
+    // 1 step gradient descent
+    optimizer.step();
+
+    // Print out the loss after 1-step gradient descent
+    pred = linear(x);
+    loss = criterion(pred, y);
+    std::cout << "Loss after 1 optimization step: " << loss.item<double>() << "\n\n";
+
+    // =============================================================== //
+    //               CREATING TENSORS FROM EXISTING DATA               //
+    // =============================================================== //
+
+    std::cout << "---- CREATING TENSORS FROM EXISTING DATA ----\n";
+
+    // WARNING: Tensors created with torch::from_blob(ptr_to_data, ...) do not own
+    // the memory pointed to by ptr_to_data!
+    // (see https://pytorch.org/cppdocs/notes/tensor_basics.html#using-externally-created-data)
+    //
+    // If you want a tensor that has its own copy of the data you can call clone() on the
+    // tensor returned from torch::from_blob(), e.g.:
+    // torch::Tensor t = torch::from_blob(ptr_to_data, ...).clone();
+    // (see https://github.com/pytorch/pytorch/issues/12506#issuecomment-429573396)
+
+    // Tensor From C-style array
+    float data_array[] = {1, 2, 3, 4};
+    torch::Tensor t1 = torch::from_blob(data_array, {2, 2});
+    std::cout << "Tensor from array:\n" << t1 << '\n';
+
+    TORCH_CHECK(data_array == t1.data_ptr<float>());
+
+    // Tensor from vector:
+    std::vector<float> data_vector = {1, 2, 3, 4};
+    torch::Tensor t2 = torch::from_blob(data_vector.data(), {2, 2});
+    std::cout << "Tensor from vector:\n" << t2 << "\n\n";
+
+    TORCH_CHECK(data_vector.data() == t2.data_ptr<float>());
+
+    // =============================================================== //
+    //                         INPUT PIPELINE                          //
+    // =============================================================== //
+
+    std::cout << "---- INPUT PIPELINE ----\n";
+
+    // Construct MNIST dataset
+    const std::string MNIST_data_path = "../../../../data/mnist/";
+
+    auto dataset = torch::data::datasets::MNIST(MNIST_data_path)
+        .map(torch::data::transforms::Normalize<>(0.1307, 0.3081))
+        .map(torch::data::transforms::Stack<>());
+
+    // Fetch one data pair
+    auto example = dataset.get_batch(0);
+    std::cout << "Sample data size: ";
+    print_tensor_size(example.data);
+    std::cout << "\n";
+    std::cout << "Sample target: " << example.target.item<int>() << "\n";
+
+    // Construct data loader
+    auto dataloader = torch::data::make_data_loader<torch::data::samplers::RandomSampler>(
+        dataset, /*batch size=*/ 64);
```

Comment:
We can skip the commenting of arguments for all dataloaders

Reply:
Agreed!